posts/2019-02-06-algebraic-api-design.md

```diff
@@ -0,0 +1,830 @@
+          .filter(!CellState.isEmpty(_))
+
+        val diff = column.cells.length - nonEmptyCells.length
+
+        Column(nonEmptyCells ++ List.fill(diff)(Empty))
+      }
+    }
+
+    def empty(height: Int): Column = Column((1 to height).map(_ => Empty).toList)
+  }
+
+  object CellState {
+    def isEmpty(cellState: CellState): Boolean = {
+      cellState match {
+        case Empty => true
+        case _     => false
+      }
+    }
+  }
+
+  object Board {
+    implicit class ColumnMapper(board: Board) {
+      def map(f: (Column, Int) => Column): Board = {
+        Board(board.columns.zipWithIndex.map { case (c, i) => f(c, i) })
+      }
+
+      def shiftLeft: Board = {
+        val nonEmptyColumns = board.columns
+          .filter(column => !CellState.isEmpty(column.cells.head))
+
+        val diff   = board.columns.length - nonEmptyColumns.length
+        val height = board.columns.head.cells.length
+        Board(nonEmptyColumns ++ List.fill(diff)(Column.empty(height)))
+      }
+    }
+  }
+
+  private val bonus = 1000
+
+  private def sqr(x: Int): Int = x * x
+
+  private def calcScore(group: Group): Int = sqr(group.positions.size - 2)
+
+  private def penalty(numberOfFilledCells: Int): Int = -sqr(numberOfFilledCells - 2)
+
+  private def getCellState(board: Board, position: Position): CellState = {
+    val width  = board.columns.length
+    val height = board.columns.head.cells.length
+    if (position.col >= 0 && position.col < width && position.row >= 0 && position.row < height) {
+      board.columns(position.col).cells(position.row)
+    } else {
+      Empty
+    }
+  }
+
+  private def findAdjacentWithSameColor(board: Board, position: Position): Set[Position] = {
+    getCellState(board, position) match {
+      case Filled(color) =>
+        Set(
+          Position.up(position),
+          Position.right(position),
+          Position.down(position),
+          Position.left(position)
+        ).map(p => (getCellState(board, p), p))
+          .filter {
+            case (Filled(c), _) => c == color
+            case _              => false
+          }
+          .map(_._2)
+
+      case Empty => Set()
+    }
+  }
+
+  private def hasValidMoves(board: Board): Boolean = {
+    board.columns.zipWithIndex
+      .exists {
+        case (column, colIndex) =>
+          column.cells.zipWithIndex
+            .exists {
+              case (_, rowIndex) =>
+                findAdjacentWithSameColor(board, Position(colIndex, rowIndex)).nonEmpty
+            }
+      }
+  }
+
+  private def filledCells(board: Board): Int = {
+    board.columns
+      .foldLeft(0)((total, column) =>
+        column.cells.foldLeft(total)((count, cell) =>
+          cell match {
+            case Filled(_) => count + 1
+            case Empty     => count
+        }))
+  }
+
+  private def findGroup(board: Board, position: Position): Option[Group] = {
+    def find(toSearch: Set[Position], group: Set[Position]): Set[Position] = {
```

Comment:
How about folding over `headOption` instead of `if/else` ? Also you could remove the extra `{}`.

```scala
toSearch.headOption.fold(group) { head =>
  val cellsWithSameColor = findAdjacentWithSameColor(board, head)
  val cellsFoundSoFar    = group + head
  val stillToSearch      = (cellsWithSameColor ++ toSearch.tail) -- cellsFoundSoFar
  find(stillToSearch, cellsFoundSoFar)
}
```